flink-java/src/main/java/org/apache/flink/api/java/operators/SortPartitionOperator.java

```diff
@@ -79,58 +119,41 @@ public SortPartitionOperator(DataSet<T> dataSet, String sortField, Order sortOrd
 	 *
 	 * @param field The field expression referring to the field of the additional sort order of
-	 *                 the local partition sorting.
-	 * @param order The order  of the additional sort order of the local partition sorting.
+	 *              the local partition sorting.
+	 * @param order The order of the additional sort order of the local partition sorting.
 	 * @return The DataSet with sorted local partitions.
 	 */
 	public SortPartitionOperator<T> sortPartition(String field, Order order) {
-		int[] flatOrderKeys = getFlatFields(field);
-		this.appendSorting(flatOrderKeys, order);
+		if (useKeySelector) {
+			throw new InvalidProgramException("Expression keys cannot be appended after a KeySelector");
+		}
+
+		ensureSortableKey(field);
+		keys.add(new Keys.ExpressionKeys<>(field, getType()));
+		orders.add(order);
+
 		return this;
 	}
 
-	// --------------------------------------------------------------------------------------------
-	//  Key Extraction
-	// --------------------------------------------------------------------------------------------
-
-	private int[] getFlatFields(int field) {
+	public <K> SortPartitionOperator<T> sortPartition(KeySelector<T, K> keyExtractor, Order order) {
+		throw new InvalidProgramException("KeySelector cannot be chained.");
+	}
 
-		if (!Keys.ExpressionKeys.isSortKey(field, super.getType())) {
+	private void ensureSortableKey(int field) throws InvalidProgramException {
+		if (!Keys.ExpressionKeys.isSortKey(field, getType())) {
 			throw new InvalidProgramException("Selected sort key is not a sortable type");
 		}
-
-		Keys.ExpressionKeys<T> ek = new Keys.ExpressionKeys<>(field, super.getType());
-		return ek.computeLogicalKeyPositions();
 	}
 
-	private int[] getFlatFields(String fields) {
-
-		if (!Keys.ExpressionKeys.isSortKey(fields, super.getType())) {
+	private void ensureSortableKey(String field) throws InvalidProgramException {
+		if (!Keys.ExpressionKeys.isSortKey(field, getType())) {
 			throw new InvalidProgramException("Selected sort key is not a sortable type");
 		}
-
-		Keys.ExpressionKeys<T> ek = new Keys.ExpressionKeys<>(fields, super.getType());
-		return ek.computeLogicalKeyPositions();
 	}
 
-	private void appendSorting(int[] flatOrderFields, Order order) {
-
-		if(this.sortKeyPositions == null) {
-			// set sorting info
-			this.sortKeyPositions = flatOrderFields;
-			this.sortOrders = new Order[flatOrderFields.length];
-			Arrays.fill(this.sortOrders, order);
-		} else {
-			// append sorting info to exising info
-			int oldLength = this.sortKeyPositions.length;
-			int newLength = oldLength + flatOrderFields.length;
-			this.sortKeyPositions = Arrays.copyOf(this.sortKeyPositions, newLength);
-			this.sortOrders = Arrays.copyOf(this.sortOrders, newLength);
-
-			for(int i=0; i<flatOrderFields.length; i++) {
-				this.sortKeyPositions[oldLength+i] = flatOrderFields[i];
-				this.sortOrders[oldLength+i] = order;
-			}
+	private void ensureSortableKey(Keys<T> sortKey) {
```

Comment:
Change the `sortKey` parameter type to `SelectorFunctionKeys` and remove the type check + cast.